packages/salesforcedx-vscode-core/src/commands/forceApexClassCreate.ts

```diff
@@ -0,0 +1,135 @@
+import {
+  CancelResponse,
+  CompositeParametersGatherer,
+  ContinueResponse,
+  DirFileNameSelection,
+  ParametersGatherer,
+  SfdxCommandlet,
+  SfdxCommandletExecutor,
+  SfdxWorkspaceChecker
+} from './commands';
+
+class SelectFileName implements ParametersGatherer<{ fileName: string }> {
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<{ fileName: string }>
+  > {
+    const fileNameInputOptions = <vscode.InputBoxOptions>{
+      prompt: nls.localize('force_apex_class_create_enter_file_name')
+    };
+    const fileName = await vscode.window.showInputBox(fileNameInputOptions);
+    return fileName
+      ? { type: 'CONTINUE', data: { fileName } }
+      : { type: 'CANCEL' };
+  }
+}
+
+class SelectDirPath implements ParametersGatherer<{ outputdir: string }> {
+  private explorerDir: string | undefined;
+
+  public constructor(explorerDir?: { path: string }) {
+    this.explorerDir = explorerDir ? explorerDir.path : explorerDir;
+  }
+
+  public globDirs(srcPath: string, prioritize?: string): string[] {
+    const unprioritizedDirs = new glob.GlobSync(srcPath + '/**/').found;
+    if (prioritize) {
+      const notPrioritized: string[] = [];
+      const prioritized = unprioritizedDirs.filter(dir => {
+        if (dir.includes(prioritize)) {
+          return true;
+        } else {
+          notPrioritized.push(dir);
+        }
+      });
+      return prioritized.concat(notPrioritized);
+    }
+    return unprioritizedDirs;
+  }
+
+  public async gather(): Promise<
+    CancelResponse | ContinueResponse<{ outputdir: string }>
+  > {
+    const rootPath = vscode.workspace.rootPath ? vscode.workspace.rootPath : '';
+    const outputdir = this.explorerDir
+      ? this.explorerDir
+      : await vscode.window.showQuickPick(
+          this.globDirs(rootPath, 'classes'),
+          <vscode.QuickPickOptions>{
+            placeHolder: nls.localize('force_apex_class_create_enter_dir_name')
+          }
+        );
+    return outputdir
+      ? { type: 'CONTINUE', data: { outputdir } }
+      : { type: 'CANCEL' };
+  }
+}
+
+class ForceApexClassCreateExecutor extends SfdxCommandletExecutor<
+  DirFileNameSelection
+> {
+  public build(data: DirFileNameSelection): Command {
+    return new SfdxCommandBuilder()
+      .withDescription(nls.localize('force_apex_class_create'))
+      .withArg('force:apex:class:create')
+      .withFlag('--classname', data.fileName)
+      .withFlag('--template', 'DefaultApexClass')
+      .withFlag('--outputdir', data.outputdir)
+      .build();
+  }
+
+  public execute(response: ContinueResponse<DirFileNameSelection>): void {
+    const cancellationTokenSource = new vscode.CancellationTokenSource();
+    const cancellationToken = cancellationTokenSource.token;
+
+    const execution = new CliCommandExecutor(this.build(response.data), {
+      cwd: vscode.workspace.rootPath
+    }).execute(cancellationToken);
+
+    execution.processExitSubject.subscribe(async data => {
```

Comment:
Rather than having a filewatcher watching for onDidCreate events, we subscribe to the exit subject of the command execution and open the file created after seeing a successful exit code.

Also removed the notificationService as discussed while going over the flow.